### PR TITLE
arch: arm: Define & implement API for test target (Non-Secure)

### PR DESCRIPTION
**Pull request by ioannisg**
This commit defines and implements an internal ARMv8-M API
that allows the user to evaluate access permissions of memory
locations, based on the ARMv8-M Test Target (TT) instruction
support.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

### CODE REVIEW

**Comment by ioannisg** [comment]
- Removing RFC label (treat as self-contained PR for the ARMv8-M TT feature)
- Verified locally

**Comment by ioannisg** [comment]
@agross-linaro take a look when you have time at this minimal API for the TT feature in ARMv8-M.
(Restricted to the parts of the feature that are generic to ARMv8-M and, thus, do not require TrustZone.)

**Comment by ioannisg** [comment]
@galak, @agross-linaro could someone from arch/arm maintainers take a look at this PR and provide some feedback?

**Comment by carlescufi** [comment]
Any progress on this one? It's been open for a month now.

File: arch/arm/core/cortex_m/cmse/arm_core_cmse.c
Comment (andrewboie) [inline arch/arm/core/cortex_m/cmse/arm_core_cmse.c]: the new standard is to prefix private kernel APIs with z_ and not _.
However, since this is a static function, the leading _ is not necessary.

**Review by tyrantlizard77 — APPROVED** [review]

**Comment by ioannisg** [comment]
Re-based (after merging #7215)

**Review by nashif — APPROVED** [review]

File: arch/arm/include/cortex_m/cmse.h
Comment (galak) [inline arch/arm/include/cortex_m/cmse.h]: This looks weird, is this all correct?

File: arch/arm/core/CMakeLists.txt
Comment (galak) [inline arch/arm/core/CMakeLists.txt]: Does this really need to be a subdir, do we expect more files in here?

**Review by galak — COMMENTED** [review]
A few inline comments/questions.

File: arch/arm/include/cortex_m/cmse.h
Comment (ioannisg) [reply arch/arm/include/cortex_m/cmse.h]: Yes, this is needed for the <arm_cmse.h> macro definitions, such as:

cmse_check_pointed_object(p, f)

File: arch/arm/core/CMakeLists.txt
Comment (ioannisg) [reply arch/arm/core/CMakeLists.txt]: @galak, since this is specific to ARMv8-M, I wouldn't prefer to "pollute" the cortex-m/ directory, which is generic to all Cortex-M  MCUs; that's why I created this new sub-directory.

It might get a Kconfig, later, too, as we will bring-in the secure part of the API, so it gets really cleaner IMO with a separate folder.

File: arch/arm/include/cortex_m/cmse.h
Comment (galak) [reply arch/arm/include/cortex_m/cmse.h]: Should this be `#ifndef typeof`

Comment (ioannisg) [reply arch/arm/include/cortex_m/cmse.h]: Ah, right, there is an ugly typo there. Will fix, thanks

File: arch/arm/core/CMakeLists.txt
Comment (carlescufi) [reply arch/arm/core/CMakeLists.txt]: I also think it make sense, especially given that there's already an `mpu` folder

File: arch/arm/core/cortex_m/cmse/arm_core_cmse.c
Comment (carlescufi) [reply arch/arm/core/cortex_m/cmse/arm_core_cmse.c]: Agreed, remove leading underscore

File: arch/arm/core/cortex_m/cmse/arm_core_cmse.c
Comment (carlescufi) [inline arch/arm/core/cortex_m/cmse/arm_core_cmse.c]: remove leading underscore

**Review by carlescufi — CHANGES_REQUESTED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**
